src/event_generator.hpp

```diff
@@ -0,0 +1,209 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <random>
+
+#include <common_types.hpp>
+#include <event_queue.hpp>
+#include <util/range.hpp>
+#include <util/rangeutil.hpp>
+
+namespace arb {
+
+struct event_generator {
+    // Return the next event
+    // Should return the same event if called multiple times without calling
+    // event_generator::pop().
+    virtual postsynaptic_spike_event next() = 0;
+    virtual void pop() = 0;
+    virtual void reset() = 0;
+    virtual ~event_generator() {};
+    virtual void advance(time_type t) = 0;
+};
+
+inline
+postsynaptic_spike_event terminal_pse() {
+    return postsynaptic_spike_event{cell_member_type{0,0}, max_time, 0};
+}
+
+// Generator that feeds events that are specified with a vector
+struct vector_backed_generator: public event_generator {
+    using pse = postsynaptic_spike_event;
+    vector_backed_generator(pse_vector events):
+        events_(std::move(events)),
+        it_(events_.begin())
+    {
+        if (!std::is_sorted(events_.begin(), events_.end())) {
+            util::sort(events_);
+        }
+    }
+
+    postsynaptic_spike_event next() override {
+        return it_==events_.end()? terminal_pse(): *it_;
+    }
+
+    void pop() override {
```

Comment:
I would expect a pop() to return something. And compared with next() it might be a little confusing

Reply:
I agree: I sometimes find it confusing that `pop` doesn't return the value at the front of the queue, but that is how I have seen it implemented in most places, and it is also the interface of `std::queue`:

http://en.cppreference.com/w/cpp/container/queue

To get the next value from the generator, do you have suggestions? `std::queue` uses `front`, and I think that `peek` is often used too.

I am sure that @halflat has an opinion.

Reply:
"Removes an element from the front of the queue. Effectively calls c.pop_front()"

Reply:
I am going with `pop`, because it is what `std::queue` and `std::priority_queue` use.
`pop_front()` is not part of the interface for the STL queue types: it is an interface that must be provided by the optional user-defined container used by the queue.